File: scheduler/sqlite-persistence/src/test/java/org/conductoross/conductor/scheduler/sqlite/dao/SqliteSchedulerArchivalDAOTest.java
package org.conductoross.conductor.scheduler.sqlite.dao;

import javax.sql.DataSource;

import org.flywaydb.core.Flyway;
import org.junit.Before;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.DependsOn;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit4.SpringRunner;

import com.netflix.conductor.common.config.ObjectMapperProvider;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.orkes.conductor.dao.archive.SchedulerArchivalDAO;
import io.orkes.conductor.scheduler.dao.AbstractSchedulerArchivalDAOTest;

/**
 * Runs the full {@link AbstractSchedulerArchivalDAOTest} contract suite against an in-memory SQLite
 * database.
 */
@ContextConfiguration(
        classes = {
            DataSourceAutoConfiguration.class,
            SqliteSchedulerArchivalDAOTest.SqliteTestConfiguration.class
        })
@RunWith(SpringRunner.class)
@SpringBootTest
@TestPropertySource(
        properties = {
            "spring.datasource.url=jdbc:sqlite::memory:",
            "spring.datasource.driver-class-name=org.sqlite.JDBC"
        })
public class SqliteSchedulerArchivalDAOTest extends AbstractSchedulerArchivalDAOTest {

    @TestConfiguration
    static class SqliteTestConfiguration {

        @Bean
        public ObjectMapper objectMapper() {
            return new ObjectMapperProvider().getObjectMapper();
        }

        @Bean(initMethod = "migrate")
        public Flyway flywayForScheduler(DataSource dataSource) {
            return Flyway.configure()
                    .locations("classpath:db/migration_scheduler_sqlite")
                    .dataSource(dataSource)
                    .table("flyway_schema_history_scheduler")
                    .outOfOrder(true)
                    .baselineOnMigrate(true)
                    .baselineVersion("0")
                    .load();
        }

        @Bean
        @DependsOn("flywayForScheduler")
        public SchedulerArchivalDAO schedulerArchivalDAO(
                DataSource dataSource, ObjectMapper objectMapper) {
            return new SqliteSchedulerArchivalDAO(dataSource, objectMapper);
        }
    }

    @Autowired private SchedulerArchivalDAO schedulerArchivalDAO;
    @Autowired private DataSource dataSource;

    @Before
    public void cleanUp() {
        JdbcTemplate jdbc = new JdbcTemplate(dataSource);
        jdbc.update("DELETE FROM workflow_scheduled_executions");
    }

    @Override
    protected SchedulerArchivalDAO archivalDao() {
        return schedulerArchivalDAO;
    }
}

File: scheduler/sqlite-persistence/src/test/typescript/org/conductoross/conductor/scheduler/sqlite/dao/SqliteSchedulerArchivalDAOTest.ts
import { describe, it, expect, beforeEach, beforeAll, afterAll } from 'vitest';
import Database from 'better-sqlite3';
import { SqliteSchedulerArchivalDAO } from '../../../../../../../../main/typescript/org/conductoross/conductor/scheduler/sqlite/dao/SqliteSchedulerArchivalDAO';
import { WorkflowScheduleExecutionModel } from '../../../../../../../../../../mock';

/**
 * Runs the full contract suite against an in-memory SQLite database.
 */
describe('SqliteSchedulerArchivalDAOTest', () => {
    let db: any;
    let schedulerArchivalDAO: SqliteSchedulerArchivalDAO;

    beforeAll(() => {
        // Equivalent to spring.datasource.url=jdbc:sqlite::memory:
        db = new Database(':memory:');
        
        // Mock flyway migration for scheduler tables
        db.exec(`
            CREATE TABLE workflow_scheduled_executions (
                execution_id TEXT PRIMARY KEY,
                schedule_name TEXT,
                workflow_name TEXT,
                workflow_id TEXT,
                reason TEXT,
                stack_trace TEXT,
                state TEXT,
                scheduled_time INTEGER,
                execution_time INTEGER,
                start_workflow_request TEXT
            );
        `);
        
        schedulerArchivalDAO = new SqliteSchedulerArchivalDAO(db);
    });

    afterAll(() => {
        if (db) {
            db.close();
        }
    });

    beforeEach(() => {
        db.exec("DELETE FROM workflow_scheduled_executions");
    });

    it('should save and read execution record', async () => {
        const executionModel = new WorkflowScheduleExecutionModel();
        executionModel.executionId = 'exec-1';
        executionModel.scheduleName = 'schedule-1';
        executionModel.workflowName = 'test-workflow';
        executionModel.scheduledTime = 1000;
        
        await schedulerArchivalDAO.saveExecutionRecord(executionModel);
        
        const retrieved = await schedulerArchivalDAO.getExecutionById('exec-1');
        expect(retrieved).not.toBeNull();
        expect(retrieved?.executionId).toBe('exec-1');
        expect(retrieved?.scheduleName).toBe('schedule-1');
        expect(retrieved?.workflowName).toBe('test-workflow');
    });

    it('should clean up old records', async () => {
        for (let i = 0; i < 5; i++) {
            const executionModel = new WorkflowScheduleExecutionModel();
            executionModel.executionId = `exec-${i}`;
            executionModel.scheduleName = 'schedule-cleanup';
            executionModel.scheduledTime = 1000 + i;
            await schedulerArchivalDAO.saveExecutionRecord(executionModel);
        }

        await schedulerArchivalDAO.cleanupOldRecords(2, 3);
        
        const result = await schedulerArchivalDAO.searchScheduledExecutions("schedule_name=schedule-cleanup", "", 0, 10, []);
        expect(result.results.length).toBe(2);
        // Ensure the remaining ones are the most recent (exec-3 and exec-4)
        expect(result.results).toContain('exec-4');
        expect(result.results).toContain('exec-3');
    });
});
